execution/gethexec/wasmstorerebuilder.go

```diff
@@ -0,0 +1,115 @@
+	}
+	return val, nil
+}
+
+func WriteToKeyValueStore[T any](store ethdb.KeyValueStore, key []byte, val T) error {
+	valBytes, err := rlp.EncodeToBytes(val)
+	if err != nil {
+		return err
+	}
+	err = store.Put(key, valBytes)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RebuildWasmStore function runs a loop looking at every codehash in diskDb, checking if its an activated stylus contract and
+// saving it to wasm store if it doesnt already exists. When errored it logs them and silently returns
+//
+// It stores the status of rebuilding to wasm store by updating the codehash (of the latest sucessfully checked contract) in
+// RebuildingPositionKey every 50 checks.
+//
+// It also stores a special value that is only set once when rebuilding commenced in RebuildingStartBlockHashKey as the block
+// time of the latest block when rebuilding was first called, this is used to avoid recomputing of assembly and module of
+// contracts that were created after rebuilding commenced since they would anyway already be added during sync.
+func RebuildWasmStore(ctx context.Context, wasmStore ethdb.KeyValueStore, l2Blockchain *core.BlockChain, position, rebuildingStartBlockHash common.Hash) {
+	var err error
+	var stateDb *state.StateDB
+	latestHeader := l2Blockchain.CurrentBlock()
+	// Attempt to get state at the start block when rebuilding commenced, if not available (in case of non-archival nodes) use latest state
+	rebuildingStartHeader := l2Blockchain.GetHeaderByHash(rebuildingStartBlockHash)
+	stateDb, err = l2Blockchain.StateAt(rebuildingStartHeader.Root)
+	if err != nil {
+		log.Info("error getting state at start block of rebuilding wasm store, attempting rebuilding with latest state", "err", err)
+		stateDb, err = l2Blockchain.StateAt(latestHeader.Root)
+		if err != nil {
+			log.Error("error getting state at latest block, aborting rebuilding", "err", err)
+			return
+		}
+	}
+	diskDb := stateDb.Database().DiskDB()
+	arbState, err := arbosState.OpenSystemArbosState(stateDb, nil, true)
+	if err != nil {
+		log.Error("error getting arbos state, aborting rebuilding", "err", err)
+		return
+	}
+	programs := arbState.Programs()
+	iter := diskDb.NewIterator(rawdb.CodePrefix, position[:])
+	for count := 1; iter.Next(); count++ {
+		codeHashBytes := bytes.TrimPrefix(iter.Key(), rawdb.CodePrefix)
+		codeHash := common.BytesToHash(codeHashBytes)
+		code := iter.Value()
+		if state.IsStylusProgram(code) {
+			if err := programs.SaveActiveProgramToWasmStore(stateDb, codeHash, code, latestHeader.Time, l2Blockchain.Config().DebugMode(), rebuildingStartHeader.Time); err != nil {
+				log.Error("error while rebuilding of wasm store, aborting rebuilding", "err", err)
+				return
+			}
+		}
+		// After every fifty codeHash checks, update the rebuilding position
+		// This also notifies user that we are working on rebuilding
+		if count%50 == 0 || ctx.Err() != nil {
```

Comment:
I was wrong to suggest every count% something.
Better would be to remember last time you did log.Info/writeKey and do another one if some time passed (1 second?)
That will give us both a constant flow of logs and make sure if the node died unexpectedly we only lose ~1second of work.

Reply:
1 second looks good, I changed the behavior